linera-views/DESIGN.md

```diff
@@ -0,0 +1,140 @@
+* `DynamoDbClient` is the AWS-based DynamoDB service.
+
+The trait `KeyValueStoreClient` was designed so that more storage solutions can be easily added in the future.
+
+The `KeyValueStoreClient` trait is also implemented for several internal constructions of clients:
+* The `LruCachingKeyValueClient` client implements the Least Recently Used (LRU)
+caching of reads into the client.
+* The `ViewContainer<C>` client implements a key-value store client from a context.
+* The `ValueSplittingKeyValueStoreClient<K>` implements a client for which the
+size of the values is unbounded, on top of another client for which it is bounded.
+(Some databases have strict limitations on the value size.)
+
+## Views
+
+A view is a container that mimics some of the properties of classic containers such
+as `VecDeque`, `BTreeMap`, etc. with the following goals:
+* The data structure can be written to the key-value store and loaded from there.
+* The changes are first done in memory and accessed there. The changes can be committed
+to the key-value store client when adequate.
+
+There are various views corresponding to several use cases. All of those are
+implemented by mapping the values in their specific types to their serializations.
+
+## Allowed keys, base_key and prefixes
+
+The keys are constructed step by step from the prefix to the full key. So, whenever
+we have a view or any other object then we have a `base_key` associated with it and
+with no other objects. A view is associated to many keys, all of whom share the same
+`base_key` as prefix.
+This leads us to introduce a trait `Context` that can be implemented with a specific
+`KeyValueStoreClient` client and a `base_key`.
+
+Another issue to consider is that we do not want to just store the values, but we
+also need to accommodate other features:
+* Administrative data (like journal) need to store some data starting from the `base_key`.
+* Some of the values need to be split into several smaller values.
+
+The rule for constructing are the following:
+* For the construction of `struct` objects of associated base key `base_key` we do the
+following: If the corresponding type has entries `entry_0`, ..., `entry_k` then the
+base key of the object associated with the `k`-th entry is `[base_key * * * *]` with `[* * * *]`
+the serialization of the number `k` considered as a `u32`.
+* For each view type, we have some associated data. Some are the hash, others the counts
+(e.g. for `LogView` and `QueueView`) and some the index. The corresponding enum is
+named `KeyTag` for each of those containers. For the index, the values are serialized.
+* For a given base key `base_key` the keys of the form `[base_key tag]` with tags in
+`0..MIN_VIEW_TAG` are reserved for administrative data.
+
+These constructions are done in order to satisfy the following property: The set of keys
+of a view is prefix-free. That is if a key is of the form `key = [a b]` then `a` is
+not a key.
```

Comment:
We also require that the user of `linera-views` makes sure the base keys they use are prefix-free, I think? We should document this anywhere where the user can explicitly assign a base key, like `ContextFromDb::create`.

Reply:
That is fair, and I will add a sentence over it.
However, if the user is using a single view then no issue occurs. If several views are used then the users must make sure that the respective base keys are not prefix-free.